diff -Nru acpica.dist/include/platform/acdrex.h acpica/include/platform/acdrex.h
--- acpica.dist/include/platform/acdrex.h	1970-01-01 01:00:00.000000000 +0100
+++ acpica/include/platform/acdrex.h	2014-09-08 00:40:32.000000000 +0100
@@ -0,0 +1,25 @@
+#ifndef __ACDREX_H__
+#define __ACDREX_H__
+
+#if _DREX_MACHINE==i386
+#define ACPI_MACHINE_WIDTH      32
+#endif
+
+#define ACPI_USE_NATIVE_DIVIDE
+
+#define ACPI_SYSTEM_XFACE
+#define ACPI_EXTERNAL_XFACE
+#define ACPI_INTERNAL_XFACE
+#define ACPI_INTERNAL_VAR_XFACE
+
+#ifdef _UKERNEL
+
+#include <uk/cdefs.h>
+
+#define AcpiOsPrintf printf
+#define AcpiOsVprintf vprintf
+#define ACPI_USE_ALTERNATE_PROTOTYPE_AcpiOsUnmapMemory
+#define AcpiOsUnmapMemory(...) 
+#endif /* _UKERNEL */
+
+#endif
diff -Nru acpica.dist/include/platform/acenv.h acpica/include/platform/acenv.h
--- acpica.dist/include/platform/acenv.h	2013-12-27 18:46:13.000000000 +0000
+++ acpica/include/platform/acenv.h	2014-09-08 00:38:01.000000000 +0100
@@ -146,7 +146,10 @@
  *
  *****************************************************************************/
 
-#if defined(_LINUX) || defined(__linux__)
+#if defined(_DREX_SOURCE)
+#include "acdrex.h"
+
+#elif defined(_LINUX) || defined(__linux__)
 #include "aclinux.h"
 
 #elif defined(_APPLE) || defined(__APPLE__)
