diff -Nru acpica.dist/include/acutils.h acpica/include/acutils.h
--- acpica.dist/include/acutils.h	2013-12-27 10:50:57.000000000 -0800
+++ acpica/include/acutils.h	2016-05-06 02:22:48.000000000 -0700
@@ -50,7 +50,8 @@
 
 /* Strings used by the disassembler and debugger resource dump routines */
 
-#if defined(ACPI_DISASSEMBLER) || defined (ACPI_DEBUGGER)
+#if defined(ACPI_DISASSEMBLER) || defined (ACPI_DEBUGGER) \
+	|| defined (ACPI_DEBUG_OUTPUT)
 
 extern const char                       *AcpiGbl_BmDecode[];
 extern const char                       *AcpiGbl_ConfigDecode[];
diff -Nru acpica.dist/include/platform/acdrex.h acpica/include/platform/acdrex.h
--- acpica.dist/include/platform/acdrex.h	1970-01-01 01:00:00.000000000 +0100
+++ acpica/include/platform/acdrex.h	2014-09-08 00:40:32.000000000 +0100
@@ -0,0 +1,33 @@
+#ifndef __ACDREX_H__
+#define __ACDREX_H__
+
+#if _DREX_MACHINE==i386
+#define ACPI_MACHINE_WIDTH      32
+#endif
+
+#define ACPI_USE_NATIVE_DIVIDE
+
+#define ACPI_SYSTEM_XFACE
+#define ACPI_EXTERNAL_XFACE
+#define ACPI_INTERNAL_XFACE
+#define ACPI_INTERNAL_VAR_XFACE
+
+#ifdef _UKERNEL
+
+#include <uk/cdefs.h>
+
+__decl_alias(AcpiOsPrintf, printf);
+__decl_alias(AcpiOsVprintf, vprintf);
+
+#else  /* !_UKERNEL */
+
+#include <sys/cdefs.h>
+
+__strong_alias(AcpiOsPrintf, printf);
+__strong_alias(AcpiOsVprintf, vprintf);
+
+#define ACPI_DEBUG_OUTPUT
+
+#endif /* _UKERNEL */
+
+#endif
diff -Nru acpica.dist/include/platform/acenv.h acpica/include/platform/acenv.h
--- acpica.dist/include/platform/acenv.h	2013-12-27 10:46:13.000000000 -0800
+++ acpica/include/platform/acenv.h	2016-05-06 02:20:39.000000000 -0700
@@ -146,7 +146,10 @@
  *
  *****************************************************************************/
 
-#if defined(_LINUX) || defined(__linux__)
+#if defined(_DREX_SOURCE)
+#include "acdrex.h"
+
+#elif defined(_LINUX) || defined(__linux__)
 #include "aclinux.h"
 
 #elif defined(_APPLE) || defined(__APPLE__)
