diff -Nru gcc-6.1.0.dist/config.sub gcc-6.1.0/config.sub
--- gcc-6.1.0.dist/config.sub	2015-12-31 21:13:28.000000000 +0000
+++ gcc-6.1.0/config.sub	2016-05-08 14:56:54.277867940 +0100
@@ -2,7 +2,7 @@
 # Configuration validation subroutine script.
 #   Copyright 1992-2016 Free Software Foundation, Inc.
 
-timestamp='2016-01-01'
+timestamp='2016-05-08'
 
 # This file is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License as published by
@@ -1395,7 +1395,7 @@
 	      | -interix* | -uwin* | -mks* | -rhapsody* | -darwin* | -opened* \
 	      | -openstep* | -oskit* | -conix* | -pw32* | -nonstopux* \
 	      | -storm-chaos* | -tops10* | -tenex* | -tops20* | -its* \
-	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
+	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* | -murgia* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
 	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* \
diff -Nru gcc-6.1.0.dist/gcc/config/i386/rtems.h gcc-6.1.0/gcc/config/i386/rtems.h
--- gcc-6.1.0.dist/gcc/config/i386/rtems.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-6.1.0/gcc/config/i386/rtems.h	2016-05-08 15:33:41.442995058 +0100
@@ -0,0 +1 @@
+
diff -Nru gcc-6.1.0.dist/gcc/config/murgia.h gcc-6.1.0/gcc/config/murgia.h
--- gcc-6.1.0.dist/gcc/config/murgia.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-6.1.0/gcc/config/murgia.h	2016-05-08 15:25:14.290965849 +0100
@@ -0,0 +1,17 @@
+#undef TARGET_MURGIA
+#define TARGET_MURGIA 1
+
+#define TARGET_OBJECT_SUFFIX ".obj"
+#define TARGET_EXECUTABLE_SUFFIX ".exe"
+
+#define LIB_SPEC "-lc -lmrg -lc -luk"
+
+#undef NO_IMPLICIT_EXTERN_C
+#define NO_IMPLICIT_EXTERN_C
+
+#undef TARGET_OS_CPP_BUILTINS
+#define TARGET_OS_CPP_BUILTINS()					 \
+  do {									 \
+    builtin_define_std ("__MURGIA__");					 \
+    builtin_assert ("system=murgia");					 \
+  } while (0)
diff -Nru gcc-6.1.0.dist/gcc/config.gcc gcc-6.1.0/gcc/config.gcc
--- gcc-6.1.0.dist/gcc/config.gcc	2016-04-11 11:14:59.000000000 +0100
+++ gcc-6.1.0/gcc/config.gcc	2016-05-08 15:56:36.132074231 +0100
@@ -899,6 +899,11 @@
     *) echo 'Unknown thread configuration for VxWorks'; exit 1 ;;
   esac
   ;;
+*-*-murgia*)
+  gas=yes
+  gnu_ld=yes
+  default_use_cxa_atexit=yes
+  ;;
 *-*-elf|arc*-*-elf*)
   # Assume that newlib is being used and so __cxa_atexit is provided.
   default_use_cxa_atexit=yes
@@ -1793,6 +1798,9 @@
 		tm_file="${tm_file} dbxcoff.h"
 	fi
 	;;
+i[34567]86-*-murgia*)
+	tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h i386/i386elf.h glibc-stdint.h murgia.h" 
+	;;
 ia64*-*-elf*)
 	tm_file="${tm_file} dbxelf.h elfos.h newlib-stdint.h ia64/sysv4.h ia64/elf.h"
 	tmake_file="ia64/t-ia64"
diff -Nru gcc-6.1.0.dist/gcc/ginclude/stddef.h gcc-6.1.0/gcc/ginclude/stddef.h
--- gcc-6.1.0.dist/gcc/ginclude/stddef.h	2016-01-04 14:30:50.000000000 +0000
+++ gcc-6.1.0/gcc/ginclude/stddef.h	2016-05-08 13:19:17.865530648 +0100
@@ -48,7 +48,7 @@
 
 /* On 4.3bsd-net2, make sure ansi.h is included, so we have
    one less case to deal with in the following.  */
-#if defined (__BSD_NET2__) || defined (____386BSD____) || (defined (__FreeBSD__) && (__FreeBSD__ < 5)) || defined(__NetBSD__)
+#if defined (__BSD_NET2__) || defined (____386BSD____) || (defined (__FreeBSD__) && (__FreeBSD__ < 5)) || defined(__NetBSD__) || defined(__MURGIA__)
 #include <machine/ansi.h>
 #endif
 /* On FreeBSD 5, machine/ansi.h does not exist anymore... */
diff -Nru gcc-6.1.0.dist/gcc/tsystem.h gcc-6.1.0/gcc/tsystem.h
--- gcc-6.1.0.dist/gcc/tsystem.h	2016-01-04 14:30:50.000000000 +0000
+++ gcc-6.1.0/gcc/tsystem.h	2016-05-08 12:25:16.453343964 +0100
@@ -99,13 +99,13 @@
 /* If these system headers do not exist, fixincludes must create them.  */
 #include <string.h>
 #include <stdlib.h>
-#include <unistd.h>
+//#include <unistd.h>
 
 /* GCC supplies this header.  */
 #include <limits.h>
 
 /* If these system headers do not exist, fixincludes must create them.  */
-#include <time.h>
+//#include <time.h>
 
 #endif /* inhibit_libc */
 
diff -Nru gcc-6.1.0.dist/libgcc/config.host gcc-6.1.0/libgcc/config.host
--- gcc-6.1.0.dist/libgcc/config.host	2016-02-26 20:02:28.000000000 +0000
+++ gcc-6.1.0/libgcc/config.host	2016-05-08 15:33:07.518993104 +0100
@@ -308,6 +308,9 @@
 *-*-vxworks*)
   tmake_file=t-vxworks
   ;;
+*-*-murgia*)
+  extra_parts="crtbegin.o crtend.o crtbeginS.o crtendS.o crtbeginT.o"
+  ;;
 *-*-elf)
   extra_parts="crtbegin.o crtend.o"
   ;;
@@ -746,6 +749,9 @@
 i[34567]86-*-interix[3-9]*)
 	tmake_file="$tmake_file i386/t-interix i386/t-chkstk"
 	;;
+i[34567]86-*-murgia*)
+	tmake_file="$tmake_file i386/t-crtstuff t-crtstuff-pic t-libgcc-pic"
+	;;
 ia64*-*-elf*)
 	extra_parts="$extra_parts crtbeginS.o crtendS.o crtfastmath.o"
 	tmake_file="ia64/t-ia64 ia64/t-ia64-elf ia64/t-eh-ia64 t-crtfm"
